ykcompile/src/lib.rs

```diff
@@ -141,17 +151,67 @@ impl Location {
         }
     }
 
-    /// If `self` is a `Register` then unwrap it, otherwise panic.
-    fn unwrap_reg(&self) -> u8 {
-        if let Location::Register(reg) = self {
-            *reg
+    /// If `self` is a `Mem` then return a mutable reference to its innards, otherwise panic.
+    fn unwrap_mem_mut(&mut self) -> &mut RegAndOffset {
+        if let Location::Mem(ro) = self {
+            ro
         } else {
-            panic!("tried to unwrap a Register location when it wasn't a Register");
+            panic!("tried to unwrap a Mem location when it wasn't a Mem");
+        }
+    }
+
+    /// Returns which register (if any) is used in addressing this location.
+    fn uses_reg(&self) -> Option<u8> {
+        match self {
+            Location::Register(reg) => Some(*reg),
```

Comment:
You can combine several of these along the lines of:

```rust
match self {
    Location::Register(reg)
  | Location::Mem(RegAndOffset { reg, .. })
  | Location::Indirect {  IndirectLoc::Register(reg), .. }
  <etc>
  => Some(*reg)
```

if it makes things a bit clearer?